src/ray/gcs/gcs_server/gcs_health_check_manager.h

```diff
@@ -0,0 +1,164 @@
+/// TODO (iycheng): Move the GcsHealthCheckManager to common and decouple it from GCS
+/// completely.
+class GcsHealthCheckManager {
+ public:
+  /// Constructor of GcsHealthCheckManager.
+  ///
+  /// \param io_service The thread where all operations in this class should run.
+  /// \param on_node_death_callback The callback function when some node is marked as
+  /// failure.
+  /// \param initial_delay_ms The delay for the first health check.
+  /// \param period_ms The interval between two health checks for the same node.
+  /// \param failure_threshold The threshold before a node will be marked as dead due to
+  /// health check failure.
+  GcsHealthCheckManager(
+      instrumented_io_context &io_service,
+      std::function<void(const NodeID &)> on_node_death_callback,
+      int64_t initial_delay_ms = RayConfig::instance().health_check_initial_delay_ms(),
+      int64_t timeout_ms = RayConfig::instance().health_check_timeout_ms(),
+      int64_t period_ms = RayConfig::instance().health_check_period_ms(),
+      int64_t failure_threshold = RayConfig::instance().health_check_failure_threshold());
+
+  ~GcsHealthCheckManager();
+
+  /// Start to track the healthiness of a node.
+  ///
+  /// \param node_id The id of the node.
+  /// \param channel The gRPC channel to the node.
+  void AddNode(const NodeID &node_id, std::shared_ptr<grpc::Channel> channel);
+
+  /// Stop tracking the healthiness of a node.
+  ///
+  /// \param node_id The id of the node to stop tracking.
+  void RemoveNode(const NodeID &node_id);
+
+  /// Return all the nodes monitored.
+  ///
+  /// \return A list of node id which are being monitored by this class.
+  std::vector<NodeID> GetAllNodes() const;
+
+ private:
+  void FailNode(const NodeID &node_id);
+
+  /// This is for testing. We'll use mock timer in gtest.
+#if defined(_TESTING_RAY_TIMER)
+  using Timer = _TESTING_RAY_TIMER;
+#else
+  using Timer = boost::asio::deadline_timer;
+#endif
+
+  /// The context for the health check. It's to support unary call.
+  /// It can be updated to support streaming call for efficiency.
+  class HealthCheckContext {
+   public:
+    HealthCheckContext(GcsHealthCheckManager *manager,
+                       std::shared_ptr<grpc::Channel> channel,
+                       NodeID node_id)
+        : manager_(manager),
+          node_id_(node_id),
+          stopped_(std::make_shared<bool>(false)),
+          timer_(manager->io_service_),
+          health_check_remaining_(manager->failure_threshold_) {
+      stub_ = grpc::health::v1::Health::NewStub(channel);
+      timer_.expires_from_now(
+          boost::posix_time::milliseconds(manager_->initial_delay_ms_));
+      timer_.async_wait([this](auto ec) {
+        if (ec != boost::asio::error::operation_aborted) {
+          StartHealthCheck();
+        }
+      });
+    }
+
+    ~HealthCheckContext() {
+      timer_.cancel();
+      if (context_ != nullptr) {
+        context_->TryCancel();
```

Comment:
What's the overhead of this call usually? Is it blocking?

Reply:
no overhead.